### PR TITLE
Add redir syntax error

### PR DESCRIPTION
**Pull request by Knulpinette**
Hello ! 
Can you check this line and tell me if it also has a leak ? It seems to be coming from env... not sure. 

`ls >> out | >> `
(basically any redirection operator not followed by an argument)

It gets me this error : 

```
=================================================================
==6679==ERROR: LeakSanitizer: detected memory leaks

Direct leak of 512 byte(s) in 32 object(s) allocated from:
    #0 0x7fc3d831f808 in __interceptor_malloc ../../../../src/libsanitizer/asan/asan_malloc_linux.cc:144
    #1 0x55f4eebd61d0 in ft_calloc (/home/knulp/42cursus/42_03_minishell/minishell+0xc1d0)
    #2 0x55f4eebd5747 in calloc_or_exit srcs/utils/error_exit.c:52
    #3 0x55f4eebd3bcd in new_env_content srcs/utils/env_list.c:11
    #4 0x55f4eebd3ddf in init_env_lst srcs/utils/env_list.c:30
    #5 0x55f4eebd47bc in init_minishell srcs/utils/minishell_utils.c:33
    #6 0x55f4eebd5a6b in main srcs/main.c:27
    #7 0x7fc3d7ff40b2 in __libc_start_main (/lib/x86_64-linux-gnu/libc.so.6+0x240b2)

Direct leak of 16 byte(s) in 1 object(s) allocated from:
    #0 0x7fc3d831f808 in __interceptor_malloc ../../../../src/libsanitizer/asan/asan_malloc_linux.cc:144
    #1 0x55f4eebd61d0 in ft_calloc (/home/knulp/42cursus/42_03_minishell/minishell+0xc1d0)
    #2 0x55f4eebd5747 in calloc_or_exit srcs/utils/error_exit.c:52
    #3 0x55f4eebd3bcd in new_env_content srcs/utils/env_list.c:11
    #4 0x55f4eebd3e72 in init_env_lst srcs/utils/env_list.c:36
    #5 0x55f4eebd47bc in init_minishell srcs/utils/minishell_utils.c:33
    #6 0x55f4eebd5a6b in main srcs/main.c:27
    #7 0x7fc3d7ff40b2 in __libc_start_main (/lib/x86_64-linux-gnu/libc.so.6+0x240b2)
```

Thanks !

### CODE REVIEW

**Comment by Knulpinette** [comment]
Also I put the bash manual answer to heredoc in the README. Turns out heredoc is exactly what << does. 
It's still a bit mysterious to me because I don't understand why, then, something like
 
```
echo << stop
heredoc> hello
heredoc> World
heredoc> stop

>
```
prints  a \newline. Since it's supposed to take what's in the heredoc until the delimiter as standard input... Gets me confused ^^
Maybe you get it.

**Comment by arieivs** [comment]
About your first comment
ls >> out | >>
doesn't give me any memory issue. Checked only with fsanitize. My executor is already quite different on redirections, they are being handled now. Yet, the error is weird, I don't see how the instruction/command ought to influence env... It's completely independent. Plus, it's comming from malloc itself. I had that actually, when I tried to make our minishell in the new mac from work for the first time. I was desesperate, thought it was something from the OS or who knows what. After a while, a good make fclean fixed it.

**Comment by Knulpinette** [comment]
OK. I'll try again tomorrow ! Well. Later today

Le mar. 19 avr. 2022 à 01:26, Sara Vieira ***@***.***> a
écrit :

File: srcs/parsing/lexer.c [hunk 1]
```diff
@@ -88,10 +93,29 @@ void	get_command_tables(t_cmd_table *cmd_table, int nb_cmds, char **instructions
 		if (cmd_table[i].nb_redirs)
 		{
 			cmd_table[i].redirs = get_redirs(instructions[i], cmd_table[i].nb_redirs);
+			j = 0;
+			while (j < cmd_table[i].nb_redirs)
+			{
+				if (is_empty(cmd_table[i].redirs[j].arg))
+					error_and_exit(REDIR_NO_ARG);
```
Comment (arieivs) [inline srcs/parsing/lexer.c]: You shouldn't exit. If this happens we still need to be on the shell. It won't run any other instruction from that line, but we still need to be on the shell.

**Comment by arieivs** [comment]
I think we need to rethink our error handling and the use of exit. I mean not sure how you are using it though. As far as I see we have three kinds of errors:
1. Print error message and proceed to the next command/instruction; The exit code is the one from the last instruction, so it may still be 0
2. Print error message and don't execute any other command on that line, wait for new instructions from user; The exit code is 1 or something of the sort
3. Print error message and exit the shell -> I think this will be really rare. Only a fail in malloc allocation or something of the sort. I'm starting to doubt if we should use it at all.

You can merge it! But ya, we should change this, and think better how to handle this honestly. All error in built-in functions are type 2, and failed opens in the executor are type 2 now as well.

**Comment by Knulpinette** [comment]
Let's ask !
Your mission today, if you accept it, at 42 :p

Le mar. 19 avr. 2022 à 01:37, Sara Vieira ***@***.***> a
écrit :

File: srcs/parsing/lexer.c [hunk 2]
```diff
@@ -88,10 +111,19 @@ void	get_command_tables(t_cmd_table *cmd_table, int nb_cmds, char **instructions
...
 		{
 			cmd_table[i].redirs = get_redirs(instructions[i], cmd_table[i].nb_redirs);
-			temp = rewrite_instruction_without_redirs(instructions[i]);
-			free(instructions[i]);
-			instructions[i] = temp;
+			j = 0;
+			while (j < cmd_table[i].nb_redirs)
+			{
+				if (is_empty(cmd_table[i].redirs[j].arg))
+					error_and_return(REDIR_NO_ARG, SYNTAX_ERROR);
```
Comment (arieivs) [inline srcs/parsing/lexer.c]: But what happens now when you run ls | >>
(or whatever that leads to this error)?
Cos the way I see it we need to stop it so to speak. Right now it just keeps on going.
And update minishell->exit_code.

**Comment by Knulpinette** [comment]
Ah yeah. You're right. Then we need an error don't execute

**Comment by arieivs** [comment]
I think it's more about how to structure it. how to make that value which you are returning (which doesn't need to be 258 btw) arrive to a point where it says, if I receive x, I should stop all the parsing

**Comment by Knulpinette** [comment]
Ah I see.
Well, actually it needs to be both : stop the parsing AND don't execute, considering how our code happens.

**Comment by Knulpinette** [comment]
Where I put the current syntax error, I guess instead I should write the error message and return (error_and_return). So parse() should return an exit code. That's probably the easiest way, to add that in - and ensure we can have a clean stop.

**Pull request merged**